lib/Sema/SemaBounds.cpp

```diff
@@ -578,6 +578,180 @@ namespace {
 
       return true;
     }
+
+
+    bool VisitImplicitCastExpr(ImplicitCastExpr *E) {
+      if (E->getCastKind() == CK_BitCast)
+        FindDisallowedFunctionPtrCasts(E);
+
+      return true;
+    }
+
+    bool VisitCStyleCastExpr(CStyleCastExpr *E) {
+      FindDisallowedFunctionPtrCasts(E);
+
+      return true;
+    }
+
+  private:
+    // Here we're examining places where a programmer has cast to a
+    // checked function pointer type, in order to make sure this cast is
+    // safe and valid.
+    // Let's term the outer value (after the cast), E, of type ToType.
+    // In the values we're examing, ToType is a ptr<> to a function type.
+    //
+    // To produce E, the programmer is performing a sequence of casts,
+    // both implicit and explicit, and perhaps this sequence includes using
+    // addr-of (&) or deref(*).
+    //
+    // We search this chain, starting at E.
+    //
+    // 1. If we come across something of ptr<> function type, then one of two things
+    //    happens:
+    //  a) this type is compatible with ToType, so then the cast is allowed.
+    //  b) this type not compatible, so we add an error about casting between incompatible
+    //     types and stop descending.
+    //    This allows calling functions with checked types, and local varibles
+    //    with checked function types.
+    //
+    // 2. If we come across a non-value-preserving cast, then we stop and error because
+    //    we are casting between incompatible types. Non-value-preserving casts include
+    //    casts that truncate values, and casts that change alignment. An LValueToRValue
+    //    cast is also non-value-preserving because it reads memory.
+    //
+    // 3. Eventually we may get to the end of the chain of casts. This could end in many
+    //    different kinds of expressions and values, but the ones we're allowing are:
+    //  a) 0 (the null pointer)
+    //  b) DeclRefs to unchecked top-level function definitons.
+    //
+    void FindDisallowedFunctionPtrCasts(CastExpr *E) {
+      // The type of the outer value
+      QualType ToType = E->getType();
+
+      // We're only looking for ptr<>s to functions
+      if (!ToType->isCheckedPointerPtrType() ||
+        !ToType->isFunctionPointerType())
+        return;
+
+      Expr *Needle = E->getSubExpr();
+      QualType NeedleTy = Needle->getType();
```

Comment:
It could before, but now we do the final type compatibility check after the loop, it's easier for it not to be.